sidechain/consensus/aura/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+		self.environment.init(header, shard)
+	}
+
+	fn proposing_remaining_duration(&self, slot_info: &SlotInfo<PB>) -> Duration {
+		proposing_remaining_duration(slot_info, its_consensus_slots::duration_now())
+	}
+}
+
+/// unit-testable remaining duration fn.
+fn proposing_remaining_duration<PB: ParentchainBlock>(
+	slot_info: &SlotInfo<PB>,
+	now: Duration,
+) -> Duration {
+	let proposing_duration = slot_info.duration.mul_f32(BLOCK_PROPOSAL_SLOT_PORTION);
+
+	let slot_remaining = slot_info.ends_at.checked_sub(now).unwrap_or_default();
+
+	std::cmp::min(slot_remaining, proposing_duration)
+}
+
+fn authorities<C, P, B>(
+	ocall_api: &C,
+	header: &B::Header,
+) -> Result<Vec<AuthorityId<P>>, ConsensusError>
+where
+	C: ValidateerFetch + GetStorageVerified,
+	P: Pair,
+	B: ParentchainBlock<Hash = H256>,
+{
+	Ok(ocall_api
+		.current_validateers(header)
+		.map_err(|e| ConsensusError::CouldNotGetAuthorities(e.to_string()))?
+		.into_iter()
+		.map(|e| AuthorityId::<P>::from_slice(e.pubkey.as_ref()))
+		.collect())
+}
+
+/// Get slot author for given block along with authorities.
+fn slot_author<P: Pair>(slot: Slot, authorities: &[AuthorityId<P>]) -> Option<&AuthorityId<P>> {
+	if authorities.is_empty() {
+		return None
+	}
+
+	let idx = *slot % (authorities.len() as u64);
+	assert!(
+		idx <= usize::MAX as u64,
+		"It is impossible to have a vector with length beyond the address space; qed",
+	);
+
+	let current_author = authorities.get(idx as usize).expect(
+		"authorities not empty; index constrained to list length;this is a valid index; qed",
+	);
+
+	Some(current_author)
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::mock::{default_header, validateer, EnvironmentMock, TestAura, SLOT_DURATION};
+	use itp_test::mock::onchain_mock::OnchainMock;
+	use itp_types::Block as ParentchainBlock;
+	use its_consensus_slots::{duration_now, PerShardSlotWorkerScheduler};
+	use sp_keyring::ed25519::Keyring;
+
+	fn get_aura(onchain_mock: OnchainMock) -> TestAura {
+		Aura::new(Keyring::Alice.pair(), onchain_mock, EnvironmentMock)
+	}
+
+	#[test]
+	fn current_authority_should_claim_slot() {
+		let authorities = vec![
+			Keyring::Bob.public().into(),
+			Keyring::Charlie.public().into(),
+			Keyring::Alice.public().into(),
+		];
+
+		let aura = get_aura(Default::default());
+
+		assert!(aura.claim_slot(&default_header(), 0.into(), &authorities).is_none());
+		assert!(aura.claim_slot(&default_header(), 1.into(), &authorities).is_none());
+		// this our authority
+		assert!(aura.claim_slot(&default_header(), 2.into(), &authorities).is_some());
+
+		assert!(aura.claim_slot(&default_header(), 3.into(), &authorities).is_none());
+		assert!(aura.claim_slot(&default_header(), 4.into(), &authorities).is_none());
+		// this our authority
+		assert!(aura.claim_slot(&default_header(), 5.into(), &authorities).is_some());
+	}
+
+	#[test]
+	fn on_slot_returns_block() {
+		let _ = env_logger::builder().is_test(true).try_init();
```

Comment:
I find the docs on this a bit lacking, usually with these logger init functions, you shouldn't be calling them multiple times per running program. But unfortunately the [example](https://docs.rs/env_logger/0.6.1/env_logger/#capturing-logs-in-tests) is not very explicit/clear on this. It does have a separate `init()` function, but I don't understand how it is called. Maybe something like [this](https://stackoverflow.com/questions/58006033/how-to-run-setup-code-before-any-tests-run-in-rust) is necessary?

Reply:
No, it is fine. I don't remember where I have read this, but this how it's recommended to use it in tests.

`_` contains an error if it has already been initialized, but there won't be any bad effects otherwise.